pandas/core/generic.py

```diff
@@ -3393,19 +3393,29 @@ def where(self, cond, other=np.nan, inplace=False, axis=None, level=None,
             else:
                 other = self._constructor(other, **self._construct_axes_dict())
 
+        if axis is None:
+            axis = 0
```

Comment:
From the comment below:

"The original behavior was to align cond and other with each Block along the 'columns' axis only when no axis parameter was passed to where (which can only happen in the case that other is also a DataFrame), but that doesn't work in some cases. The correct behavior seems to be: always align cond (because each Block only contains a subset of the columns), and align other whenever it has the same ndim as self, or if the alignment axis is 'columns': basically, every case except DataFrame.where(cond, Series, axis='index')

This is all fixed in the most recent commit."